### ISSUE TITLE
Fully serializable config

### ISSUE THREAD

**Issue by amannn** [amannn/next-intl#611]
### Is your feature request related to a problem? Please describe.

These three config options accept functions and can therefore not be serialized currently:
1. [`defaultTranslationValues`](https://next-intl-docs.vercel.app/docs/configuration#default-translation-values)
2. [`onError`](https://next-intl-docs.vercel.app/docs/configuration#error-handling)
3. [`getMessageFallback`](https://next-intl-docs.vercel.app/docs/configuration#error-handling)

Due to this, [we can't automatically pass them from an RSC render to the client side](https://next-intl-docs.vercel.app/docs/environments/server-client-components#non-serializable-props).

It might be worth investigating serializable alternatives, to avoid this.

### Describe the solution you'd like

**`defaultTranslationValues`**

Maybe a better option could be to deprecate the option altogether and ask users to share values e.g. via a component:

```tsx
function Markup() {
  const t = useTranslations('Markup');
  return <RichText>{components => t('markup', components)}</RichText>
}

function RichText({children}) {
  return children({
    b: (chunks: ReactNode) => <b>{chunks}</b>
  });
}
```

… or imports:

```tsx
// defaultRichTextComponents.tsx

export default {
  b: (chunks: ReactNode) => <b>{chunks}</b>
}

// Markup.tsx

import defaultRichTextComponents from './defaultRichTextComponents';

function Markup() {
  const t = useTranslations('Markup');
  return t.rich('markup', defaultRichTextComponents);
}
```

As a nice side effect, this would allow us to statically type the arguments of a given ICU string (see https://github.com/amannn/next-intl/issues/410), since there are no potential global values floating around.

Another benefit could be to remove ambiguity between values that should be used for `t.rich` and `t.markup` (see https://github.com/amannn/next-intl/issues/1181).

**`onError`**

See https://github.com/amannn/next-intl/issues/1285

**`getMessageFallback`**

See https://github.com/amannn/next-intl/issues/1285

Alternatives:
 - An ICU message like `MISSING: {namespace}.{key}`?
 - Deprecate the feature completely?

### Describe alternatives you've considered

`NextIntlClientProvider` can be added in a client-side module graph to configure the non-serializable config options, see [the docs](https://next-intl-docs-git-feat-next-13-rsc-next-intl.vercel.app/docs/environments/server-client-components#non-serializable-props).

**Comment by ovflowd** [comment]
Note this doesn't handle the case where I want to pass the messages as an object from another component.

I.e.

```tsx
const MyComponent = ({ translationValues }) => t.rich('translationKey', translatioNValues);
```

So IMHO the whole RichTranslations and MarkupTranslations should support non-functions.

**Comment by amannn** [comment]
You mean when `translationValues` contains functions, right? Can you share more about the practical use case?

**Comment by ovflowd** [comment]
> You mean when `translationValues` contains functions, right? Can you share more about the practical use case?

Wdym by that? I should be able to pass translation values as props between any component if I wish. Shouldnt be blocked by an implementation detail of making them a function.

In other words, what if I have dynamic values and want to map them, I'd be unable due to next-intl requiring markup or rich text to be wrapped on functions...

**Comment by ovflowd** [comment]
I think sharing [this](https://github.com/nodejs/nodejs.org/blob/main/layouts/DocsLayout.tsx) with you is self-explanatory.

Pretty much there's a translation key that we want to have a specific part of it to be surrounded with a "span" element (https://github.com/nodejs/nodejs.org/blob/main/i18n/locales/en.json#L84), that's it.

This was possible before, because react-intl allowed me to replace a variable, such as `{something}` (https://github.com/nodejs/nodejs.org/blob/c72dc046dffda7d8b056623602c80bac3dd6a8f4/i18n/locales/en.json#L19) with something else (https://github.com/nodejs/nodejs.org/blob/c72dc046dffda7d8b056623602c80bac3dd6a8f4/layouts/DocsLayout.tsx#L21)

I understand that this is unrelated to next-intl, and react-intl would face the same issues once I've used server components; Hence me indulging if it makes sense to support something else.

In the end, I just want to be able to share translation values from Component A to Component B. Right not it works because both components are client components, but I feel that maybe next-intl should support a way to transfer rich-text content independent if the component is RSC or CSC... 🤔 

Not a big deal, but I feel that it'd be nice if it was supported. Either by something like:

```tsx
const translationContext: RichTranslationValues = {
 graySpan: '<span className="small color-lightgray">{0}</span>'
}
```

Encapsulating the HTML code within a string and then eval'ing it. And the translation key be like:

```json
"apiLts": "{fullLtsNodeVersion} API <graySpan>something</graySpan>",
```

I know this is just a poor example, and definitely a stretch.

**Comment by ovflowd** [comment]
Also sorry if I'm just crying wold here, @amannn, please feel more than free to just disregard my request!

**Comment by amannn** [comment]
> This was possible before, because react-intl allowed me to replace a variable, such as `{something}` (https://github.com/nodejs/nodejs.org/blob/c72dc046dffda7d8b056623602c80bac3dd6a8f4/i18n/locales/en.json#L19) with something else (https://github.com/nodejs/nodejs.org/blob/c72dc046dffda7d8b056623602c80bac3dd6a8f4/layouts/DocsLayout.tsx#L21)

That's an interesting one, because you didn't need chunks previously (since "LTS" was hardcoded):

```
spanLts: <span className="small color-lightgray">LTS</span>,
```

 `next-intl` could most likely support passing a React Element without chunks for rich text, by directly assigning the element to a value (without a function). In fact, [I think it already does](https://codesandbox.io/s/use-intl-cra-example-forked-rvg63r?file=/src/App.tsx), we'd only have to update the types for such a call.

There's currently an expandable section in [the rich text docs](https://next-intl-docs.vercel.app/docs/usage/messages#rich-text) about self-closing tags. This could be updated accordingly as part of this change.

In regard to rich text content with chunks: I don't see eval'ing as a good direction here, to be honest. The provided elements can not just be HTML, but any React component.

To come back to your example: I think there could also be a question here about which parts of the component should be Server or Client Components. Briefly looking into the code, I'm wondering if the rendering of the sidebar could be achieved in RSC and only individual items that need to be highlighted if they're active could be Client Components. The next-intl App Router example [does just that](https://github.com/amannn/next-intl/blob/main/examples/example-app-router/src/components/Navigation.tsx) for the navigation.

Hope this helps!

**Comment by ovflowd** [comment]
> 'm wondering if the rendering of the sidebar could be achieved in RSC and only individual items that need to be highlighted if they're active could be Client Components.

Hm, true, I think I can make it RSC compatible.

> In regard to rich text content with chunks: I don't see eval'ing as a good direction here, to be honest. The provided elements can not just be HTML, but any React component.

Right, I'm also not a fan of this. But I believe at least for the case you mentioned above about without chunks, worth documenting it and updating the types. Can be useful for some folks :)

**Comment by amannn** [comment]
> Right, I'm also not a fan of this. But I believe at least for the case you mentioned above about without chunks, worth documenting it and updating the types. Can be useful for some folks :)

Yep, I agree!

**Comment by ovflowd** [comment]
Aaand, found a way to make it RSC by removing the only Hook that was "client-dependent"

**Comment by amannn** [comment]
Awesome! 👏

**Comment by amannn** [comment]
I've added https://github.com/amannn/next-intl/issues/1285 to investigate a better way to achieve global error handling.

**Comment by amannn** [comment]
`defaultTranslationValues` have been [deprecated](https://next-intl-docs.vercel.app/blog/next-intl-3-22#default-translation-values) in next-intl 3.22 and we suggest a better alternative now. Additionally, the error handling topic is tracked in a separate issue now: https://github.com/amannn/next-intl/issues/1285.

Due to this, I'll close this issue.
